universal-application-tool-0.0.1/app/controllers/applicant/ApplicantProgramsController.java

```diff
@@ -77,6 +84,43 @@ public ApplicantProgramsController(
             });
   }
 
+  @Secure
+  public CompletionStage<Result> view(Request request, long applicantId, long programId) {
+    CompletionStage<String> applicantStage = this.applicantService.getName(applicantId);
+
+    return applicantStage
+        .thenComposeAsync(v -> checkApplicantAuthorization(profileUtils, request, applicantId))
+        .thenComposeAsync(
+            v -> applicantService.relevantPrograms(applicantId), httpContext.current())
+        .thenApplyAsync(
+            programs -> {
+              ImmutableList<ProgramDefinition> filteredPrograms =
+                  programs.stream()
+                      .filter(program -> program.id() == programId)
+                      .collect(toImmutableList());
```

Comment:
(blocking) use `.findFirst()` instead of `collect(toImmutableList())`.

That'll give you `Optional<ProgramDefinition>`. You could reallllly do all of it with one long call

```
return programs.stream()
    .filter(program -> program.id() == programId)
    .findFirst()
    .map(programDefinition-> ok(programInfoView.render(..., ProgramDefinition, ...)))
    .orElse(badRequest());
```

Reply:
Done.